store/errors.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package store
+
+import (
+	"fmt"
+)
+
+// ErrInvalidInput indicates an error that has occured due to an invalid input.
+type ErrInvalidInput struct {
+	Entity string      // The entity which was sent as the input.
+	Field  string      // The field of the entity which was invalid.
+	Value  interface{} // The actual value of the field.
+}
+
+func NewErrInvalidInput(entity, field string, value interface{}) *ErrInvalidInput {
+	return &ErrInvalidInput{
+		Entity: entity,
+		Field:  field,
+		Value:  value,
+	}
+}
+
+func (e *ErrInvalidInput) Error() string {
+	return fmt.Sprintf("invalid input: entity: %s field: %s value: %s", e.Entity, e.Field, e.Value)
+}
+
+// ErrLimitExceeded indicates an error that has occured because some value exceeded a limit.
+type ErrLimitExceeded struct {
+	What  string // What was the object that exceeded.
+	Count int    // The value of the object.
+	meta  string // Any additional metadata.
```

Comment:
What's an example of `meta` being used? I'm wondering if it makes more sense as a `string` or as a `map[string]interface{}`.

Reply:
That's a good point. I was trying to keep it consistent with AppError, because this is how it initially looked: `model.NewAppError("SqlChannelStore.Save", "store.sql_channel.save_channel.limit.app_error", nil, "teamId="+channel.TeamId, http.StatusBadRequest)`. 

Is there a reason why we are not passing the `teamId` in the `params` map, but rather in the `details` string ?

I am totally open to changing this.

Reply:
> Is there a reason why we are not passing the teamId in the params map, but rather in the details string ?

I have no clue, I don't know the history. I guess this can come up organically in the future if necessary.